airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/cdc/PostgresCdcCtidInitializer.java

```diff
@@ -0,0 +1,181 @@
+import java.util.OptionalInt;
+import java.util.OptionalLong;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class PostgresCdcCtidInitializer {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(PostgresCdcCtidInitializer.class);
+
+  public static List<AutoCloseableIterator<AirbyteMessage>> cdcCtidIteratorsCombined(final JdbcDatabase database,
+      final ConfiguredAirbyteCatalog catalog,
+      final Map<String, TableInfo<CommonField<PostgresType>>> tableNameToTable,
+      final StateManager stateManager,
+      final Instant emittedAt,
+      final String quoteString,
+      final JsonNode replicationSlot) {
+    try {
+      final JsonNode sourceConfig = database.getSourceConfig();
+      final Duration firstRecordWaitTime = PostgresUtils.getFirstRecordWaitTime(sourceConfig);
+      final OptionalInt queueSize = OptionalInt.of(PostgresUtils.getQueueSize(sourceConfig));
+      LOGGER.info("First record waiting time: {} seconds", firstRecordWaitTime.getSeconds());
+      LOGGER.info("Queue size: {}", queueSize.getAsInt());
+
+      final PostgresDebeziumStateUtil postgresDebeziumStateUtil = new PostgresDebeziumStateUtil();
+
+      final JsonNode initialDebeziumState = postgresDebeziumStateUtil.constructInitialDebeziumState(database,
+          sourceConfig.get(JdbcUtils.DATABASE_KEY).asText());
+
+      final JsonNode state =
+          (stateManager.getCdcStateManager().getCdcState() == null || stateManager.getCdcStateManager().getCdcState().getState() == null)
+              ? initialDebeziumState
+              : Jsons.clone(stateManager.getCdcStateManager().getCdcState().getState());
+
+      final OptionalLong savedOffset = postgresDebeziumStateUtil.savedOffset(
+          Jsons.clone(PostgresCdcProperties.getDebeziumDefaultProperties(database)),
+          catalog,
+          state,
+          sourceConfig);
+
+      // We should always be able to extract offset out of state if it's not null
+      if (state != null && savedOffset.isEmpty()) {
+        throw new RuntimeException(
+            "Unable extract the offset out of state, State mutation might not be working. " + state.asText());
+      }
+
+      final boolean savedOffsetAfterReplicationSlotLSN = postgresDebeziumStateUtil.isSavedOffsetAfterReplicationSlotLSN(
+          // We can assume that there will be only 1 replication slot cause before the sync starts for
+          // Postgres CDC,
+          // we run all the check operations and one of the check validates that the replication slot exists
+          // and has only 1 entry
+          replicationSlot,
+          savedOffset);
+
+      if (!savedOffsetAfterReplicationSlotLSN) {
+        LOGGER.warn("Saved offset is before Replication slot's confirmed_flush_lsn, Airbyte will trigger sync from scratch");
+      } else if (PostgresUtils.shouldFlushAfterSync(sourceConfig)) {
```

Comment:
I know (think?) this is a refactor,
But can you please explain this line? 
When is `savedOffsetAfterReplicationSlotLSN == true` and how does it relate to `PostgresUtils.shouldFlushAfterSync()`?

Reply:
The `savedOffsetAfterReplicationSlotLSN` parameter denotes whether the saved LSN in our state is ahead of the `confirmed_flush_lsn` attribute of the replication slot. This can be true if the user has selected the option of acknowledging the LSN while reading WAL during the setup of the connector and then a failure happened in middle. This can also be true if another process (not Airbyte) is using the same replication slot and acknowledged the lsn.

PostgresUtils.shouldFlushAfterSync() is the second behaviour that you can select on how Airbyte should ack the lsn. This is the default i.e. we ack what we have in our state at the beginning of the sync.